### PR TITLE
T6949: adds blackbox exporter

### PR DESCRIPTION
**Pull request by nvollmar**
## Change Summary
Adds Prometheus Blackbox Exporter

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
* https://vyos.dev/T6949

## Related PR(s)
* vyos/vyos-build#871
* vyos/vyos-documentation#1582

## Component(s) name
monitoring

## Proposed changes
Adding Blackbox Exporter allows Prometheus to probe endpoints over HTTP, HTTPS, DNS, TCP, ICMP and gRPC.
This can deliver metrics about upstream connectivity, dns service responsiveness etc.

## How to test
Enable blackbox_exporter
```
set service monitoring prometheus blackbox-exporter listen-address 0.0.0.0
```
Then check service is running
```
curl localhost:9115
```

## Smoketest result
```
vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_monitoring_prometheus.py
test_01_node_exporter (__main__.TestMonitoringPrometheus.test_01_node_exporter) ... ok
test_02_frr_exporter (__main__.TestMonitoringPrometheus.test_02_frr_exporter) ... ok
test_03_blackbox_exporter (__main__.TestMonitoringPrometheus.test_03_blackbox_exporter) ... ok
test_04_blackbox_exporter_with_config (__main__.TestMonitoringPrometheus.test_04_blackbox_exporter_with_config) ... ok
```

## Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [X] I have linked this PR to one or more Phabricator Task(s)
- [X] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [X] My commit headlines contain a valid Task id
- [X] My change requires a change to the documentation
- [X] I have updated the documentation accordingly

### CODE REVIEW

File: interface-definitions/service_monitoring_prometheus.xml.in [hunk 1]
```diff
@@ -36,6 +36,31 @@
...
+                  <help>Prometheus exporter for probing endpoints</help>
+                </properties>
+                <children>
+                  #include <include/listen-address.xml.i>
+                  #include <include/port-number.xml.i>
+                  <leafNode name="port">
+                    <defaultValue>9115</defaultValue>
+                  </leafNode>
+                  #include <include/interface/vrf.xml.i>
+                  <leafNode name="config-file">
```
Comment (c-po) [inline interface-definitions/service_monitoring_prometheus.xml.in]: I am no fan at all of out-of CLI configuration files / snippets. Those usually do more harm then good.
We should rather add the required CLI knows instead.

File: src/conf_mode/service_monitoring_prometheus.py [hunk 2]
```diff
@@ -57,6 +60,12 @@ def get_config(config=None):
     if tmp:
         monitoring.update({'frr_exporter_restart_required': {}})
 
+    tmp = is_node_changed(conf, base + ['blackbox-exporter', 'vrf']) or is_node_changed(
```
Comment (c-po) [inline src/conf_mode/service_monitoring_prometheus.py]: Please consolidate this into a for loop
```suggestion
    for node in ['vrf', 'config-file']:
        tmp = is_node_changed(conf, base + ['blackbox-exporter', node])
```

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/service_monitoring_prometheus.xml.in (hunk 1 above)
Comment (nvollmar) [reply interface-definitions/service_monitoring_prometheus.xml.in]: I've added basic config generation for dns/icmp which are most likely used for testing upstream connectivity

File: interface-definitions/service_monitoring_prometheus.xml.in [hunk 3]
```diff
@@ -36,6 +36,137 @@
...
+                        <children>
+                          <tagNode name="name">
+                            <properties>
+                              <help>Name of the dns module</help>
+                            </properties>
+                            <children>
+                              <leafNode name="query-type">
+                                <properties>
+                                  <help>DNS query type</help>
+                                  <constraint>
```
Comment (c-po) [inline interface-definitions/service_monitoring_prometheus.xml.in:68]: Please add valueHelp XML definitions for A, AAAA and ANY - Can I also test TXT or NS?

File: interface-definitions/service_monitoring_prometheus.xml.in [hunk 4]
```diff
@@ -36,6 +36,137 @@
...
+                                  </constraint>
+                                  <constraintErrorMessage>Timeout must be between 1 and 60 seconds</constraintErrorMessage>
+                                </properties>
+                                <defaultValue>5</defaultValue>
+                              </leafNode>
+                              <leafNode name="preferred-ip-protocol">
+                                <properties>
+                                  <help>Preferred IP protocol for this module</help>
+                                  <valueHelp>
+                                    <format>ip4</format>
```
Comment (c-po) [inline interface-definitions/service_monitoring_prometheus.xml.in]: We stick with ipv4 and ipv6 literals.

File: interface-definitions/service_monitoring_prometheus.xml.in [hunk 5]
```diff
@@ -36,6 +36,137 @@
...
+                                    <description>Timeout in seconds</description>
+                                  </valueHelp>
+                                  <constraint>
+                                    <validator name="numeric" argument="--range 1-60"/>
+                                  </constraint>
+                                  <constraintErrorMessage>Timeout must be between 1 and 60 seconds</constraintErrorMessage>
+                                </properties>
+                                <defaultValue>5</defaultValue>
+                              </leafNode>
+                              <leafNode name="preferred-ip-protocol">
```
Comment (c-po) [inline interface-definitions/service_monitoring_prometheus.xml.in]: Looks like a XML definitions that's used two times for blackbox exporter.
That would suffice for an includable and reusable XML building block.

File: interface-definitions/service_monitoring_prometheus.xml.in [hunk 6]
```diff
@@ -36,6 +36,137 @@
...
+                                    <description>Prefer IPv6</description>
+                                  </valueHelp>
+                                  <constraint>
+                                    <regex>(ip4|ip6)</regex>
+                                  </constraint>
+                                </properties>
+                                <defaultValue>ip6</defaultValue>
+                              </leafNode>
+                              <leafNode name="ip-protocol-fallback">
+                                <properties>
```
Comment (c-po) [inline interface-definitions/service_monitoring_prometheus.xml.in]: Looks like a XML definitions that's used two times for blackbox exporter.
That would suffice for an includable and reusable XML building block.

File: src/conf_mode/service_monitoring_prometheus.py [hunk 7]
```diff
@@ -103,6 +120,20 @@ def generate(monitoring):
...
+    if 'blackbox_exporter' in monitoring:
+        # Render blackbox_exporter service_file
+        render(
+            blackbox_exporter_service_file,
+            'prometheus/blackbox_exporter.service.j2',
+            monitoring['blackbox_exporter'],
+        )
+        # Render blackbox_exporter config file
+        render(
+            '/etc/blackbox_exporter/config.yml',
```
Comment (c-po) [inline src/conf_mode/service_monitoring_prometheus.py]: Please render to /run

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/service_monitoring_prometheus.xml.in (hunk 3 above)
Comment (nvollmar) [reply interface-definitions/service_monitoring_prometheus.xml.in:68]: Any kind of DNS record can be requested

File: interface-definitions/service_monitoring_prometheus.xml.in [hunk 8]
```diff
@@ -36,6 +36,73 @@
...
+                                  <help>Name to be queried</help>
+                                  <constraint>
+                                    <validator name="fqdn"/>
+                                  </constraint>
+                                </properties>
+                              </leafNode>
+                              <leafNode name="query-type">
+                                <properties>
+                                  <help>DNS query type</help>
+                                  <constraint>
```
Comment (c-po) [inline interface-definitions/service_monitoring_prometheus.xml.in]:
```suggestion
                                  <valueHelp>
                                    <format>A</format>
                                    <description>DNS A record</description>
                                  </valueHelp>
                                  <valueHelp>
                                    <format>AAAA</format>
                                    <description>DNS AAAA record</description>
                                  </valueHelp>
                                  <valueHelp>
                                    <format>ANY</format>
                                    <description>Any DNS record</description>
                                  </valueHelp>                                                                    
                                  <constraint>
```

**Review by c-po — CHANGES_REQUESTED** [review]

**Review by c-po — APPROVED** [review]

**Comment by nvollmar** [comment]
The smoke tests won't work until the vyos-build changes are in the image.

**Comment by sever-sever** [comment]
Do we need to update debian/control?

**Comment by nvollmar** [comment]
> Do we need to update debian/control?

Aadded

**Review by sever-sever — APPROVED** [review]

**Comment by sever-sever** [comment]
Integration failed by not this feature

**Pull request merged**